### DISCUSSION TITLE
How to obtain a variability=tunable parameter when exporting FMU?

### DISCUSSION THREAD

**Discussion by icupeiro** [OpenModelica/OpenModelica#8781] [Q&A] [answered]
Hi all,

I want to export a model in OpenModelica to FMU including a 'tunable' input parameter to the parameter. For this, I call

` translateModelFMU(model)`

And in the model I have a set of parameters such as

`parameter input Real param;`

The resulting FMU modelDescription.xml lists the variable variability as:
 
```
<ScalarVariable
    name="param"
    valueReference="37"
    variability="fixed"
    causality="input" >
    <Real start="0.0"/>
  </ScalarVariable>
```

i.e., as a fixed input parameter. I would like to change the Modelica code to obtain a tunable parameter in the resulting FMU instead of fixed.

In Dymola, adding the annotation (Evaluate=false) to the definition of the parameter will do the trick. However, this solution seems to be specific to Dymola. I tried to add the annotation in OM but seems to have no effect regarding the resulting FMU. What would be the way to achieve this in OpenModelica?

Thanks!

**Comment by casella** [comment, marked as answer]
There is an interesting debate on this topic going on in [MLS ticket #3140](https://github.com/modelica/ModelicaSpecification/issues/3140#issuecomment-1082000359). Apparently there is some need of clarification in the Modelica and/or FMI specification. Please refer to that ticket.

**Comment by IkerLandaa** [comment]
@icupeiro I am currently trying the same, following the workflow described in an OM forum post (https://openmodelica.org/forum/default-topic/3444-how-to-obtain-a-variability-tunable-parameter-when-exporting-fmu) 

To perform this, the modelDescription.xml is changed manually to `variability="tunable"`, and then the FMU zipped again with the changes. 

I wanted to know if you followed this workflow or if there is already a solution from OM side.

In my case, executing `translateModelFMU(modelName)` does not produce the mentioned XML file, or at least I cannot find it. 

Im using OpenModelica 1.21.0 on Windows.

**Reply by icupeiro** [reply]
Hi @IkerLandaa 

I finally disregarded this track and I did not further explore on achieving tunable parameters. For my own purposes I finally manually redeclared the parameters of my interest to inputs.

Regarding updates from OM, I am not aware of those. Perhaps @casella has a better view on this

All the best,

**Reply by awecefil** [reply]
Hi @IkerLandaa , could you tell me what does `then the FMU zipped again with the changes` mean? 

Currently, I build a model in OpenModelica and then export it to FMU. OM create 2 directories, one contains model.bat, model.exe, model.log... and another contains model.fmu and other files.

After obtain the model.fmu file, I load it by FMI.jl (a julia package that support importing and manipulating FMU) and it creates a directory with the binary code, c code, and modelDescription.xml in it. If I want to follow the workflow as you mentioned in that OM forum post, I should first modify the .xml file but what should I do for the next step?

Hi @icupeiro , I also want to know how you do to manually re-declared the parameters if you don't follow that workflow?

**Reply by IkerLandaa** [reply]
Hi @awecefil , this workflow finally did not work since I was not able to zip (or rebuild) the FMU file from the folder resulting from unzipping the FMU using 7-zip.

My solution was to modify the parameters values on the OM model using OMPython and then export it to FMU after each value change. This solution takes more time than using the OMPython API so I finally discarded this solution.

**Reply by IkerLandaa** [reply]
Hi @awecefil , just to add that a colleague from my lab was finally able to tune the parameters using Julia with FMI.jl package. It may be a problem with its execution in Python, where I was using pyFMI package.

**Reply by awecefil** [reply]
Hi @IkerLandaa , I am also using FMI.jl to manipulate FMU in Julia and it really can tune the parameters "before the simulation start" but it can't makes the parameter be changed during simulation I think (from page 49 of FMI specification 2.0.4 document, it explains the difference between `fixed` and `tunable` parameter). So, there still some differences, but anyway, thanks for you help!

**Comment by javiarrobas** [comment]
I am also seeking the possibility of setting parameters as tunable for parameter estimation using the FMU. I understand that's not the conventional way of using Modelica, but ideally the OMC could expose a flag letting that happen. If there is some guidance I'm willing to invest time in implementing such a feature.

**Reply by AnHeuermann** [reply]
Contributions are always welcome! If you are willing to spend some time we can point you to the right locations in the compiler.

What information do you need?
We can point you to the locations where you can add new compiler flags and where we generate the modelDescription XML.

  - Adding a new compiler flag: [Flags.mo](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/Compiler/Util/Flags.mo#L32-L55)
  - Code generation for the modelDescription.xml is done from the Susan templates: [CodegenFMU.tpl](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/Compiler/Template/CodegenFMU.tpl)

**Reply by AnHeuermann** [reply]
But judging from https://github.com/modelica/ModelicaSpecification/issues/3140 implementing an algorithm that checks if a variable satisfy the (unclear) conditions to be tunable could be very difficult, though.

**Reply by javiarrobas** [reply]
I think what is needed is a configuration flag as defined in [Flags.mo](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/Compiler/Util/Flags.mo#L32-L55) that changes a parameter type to become a Real instead. I'd guess CodegenFMU.tpl does not need to be modified since the flag affects the compilation process, which is just downstreamed to the code generation of the FMU. 
I see that in [Flags.mo](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/Compiler/Util/Flags.mo#L32-L55) it is defined the flag and a small description is given, but the functionality is not defined there, right? Could you give a pointer to a location in the compiler where a flag functionality is defined? ideally for an example flag that does something similar to what's intended here like changing a variable type (not sure if there is one already doing something like that). 
Also, what would be the fastest workflow to check the functionality of a flag? will I need to rebuild the whole OMC every time I make an edit? I'm guessing it may be needed... but just asking in case you know a quicker way of doing so. 
By the way, if you're attending the Modelica conference this week we could have a quick meeting about this. That'd be great!
Once again, thanks a lot for your help :)

**Reply by arun3688** [reply]
we do not support the feature `variability=tunable` when exporting fmu's in OpenModelica. this is function in `SimCodeUtil.mo ` gets the variability attribute for fmus

```
protected function getVariabilityAttribute
  "function which sets the Variabilty attribute for FMI2.0 export"
  input BackendDAE.Var var;
  output SimCodeVar.Variability variability;
algorithm
  if BackendVariable.isParam(var) then
    variability := SimCodeVar.FIXED();
  elseif BackendVariable.isConst(var) then
    variability := SimCodeVar.CONSTANT();
  elseif BackendVariable.isVarDiscrete(var) or BackendVariable.isClockedStateVar(var) then
    variability := SimCodeVar.DISCRETE();
  // TODO SimCodeVar.TUNABLE()
  else
    variability := SimCodeVar.CONTINUOUS();
  end if;
end getVariabilityAttribute;
```

**Comment by casella** [comment]
Discussed this with @JavierArroyoBastida about how to proceed. Best option: introduce a vendor-specific annotation to indicate that a certain parameter has a certain variability in the FMU. Then, propagate this information to the FMU code generation.

@AnHeuermann can you provide some pointers?

**Reply by icupeiro** [reply]
@perost your colleague @AnHeuermann mentioned at the Modelica Conference that you may be able to help us out

This Modelica code represents a minimum working example 

```
model ExampleTunable
  Modelica.Blocks.Interfaces.RealInput u;

  Modelica.Thermal.HeatTransfer.Components.HeatCapacitor heatCapacitor(C=C);
  Modelica.Thermal.HeatTransfer.Components.ThermalResistor thermalResistor(R=R);
  Modelica.Thermal.HeatTransfer.Sources.PrescribedHeatFlow prescribedHeatFlow;
    
  parameter Modelica.Units.SI.HeatCapacity C = 1e8;
  parameter Modelica.Units.SI.ThermalResistance R = 1e-3;  
  
equation
  connect(u, prescribedHeatFlow.Q_flow);
  connect(prescribedHeatFlow.port, thermalResistor.port_a);
  connect(thermalResistor.port_b, heatCapacitor.port);
  annotation(
    uses(Modelica(version = "4.0.0")));
end ExampleTunable;
```

The model has one input and two parameters.
We would like to be able to:

- Change the parameters in the run time
- Get the derivative information against the parameter

This should be possible by exporting the parameter as 'tunable' in the FMU.
The trick used now in DaeBuilder is to redeclare the parameters as inputs. However, this could incur in changing some libraries base components. In this example, indeed the parameters could be redeclared as inputs. However the MSL 4.0 thermalResistor and heatCapacitor components have these declared already as parameters (thermalResistor.R and heatCapacitor.C).

Ideally, having a flag/annotation to tell the compiler that this parameter is input/tunable

Thanks a lot!

**Comment by arun3688** [comment]
> Discussed this with @JavierArroyoBastida about how to proceed. Best option: introduce a vendor-specific annotation to indicate that a certain parameter has a certain variability in the FMU. Then, propagate this information to the FMU code generation.
> 
> @AnHeuermann can you provide some pointers?

@casella If we use vendor specific annotation for making the `variablitly = tunable` then it should be much easier to implement and export to the fmus, During SimCode Phase when getting the `variablity` attribute We can check for this annotation and assign the `variability = tunable`

**Reply by casella** [reply]
Sounds good, please go ahead with that
